services/src/main/java/org/keycloak/protocol/oidc/grants/JWTBearerAssertionGrantType.java

```diff
@@ -0,0 +1,341 @@
+     * Throws an HTTP internal server error if the client configurations could not be retrieved.
+     */
+    private void checkTrustedIssuerConfig(){
+        // ensure the trusted issuer configs could be parsed
+        if (trustedIssuers == null){
+            event.detail(Details.REASON, "Unable to parse client assertion grant configuration");
+            event.error(Errors.INVALID_CONFIG);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.SERVER_ERROR, "Invalid request, unable to parse client assertion grant configuration", Response.Status.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    /**
+     * Verify the client has permissions to impersonate the user specified in the assertion JWT
+     * Throws an HTTP forbidden error if the client is not allowed to impersonate the user
+     *
+     * @param user The user in question
+     */
+    private void checkImpersonationAllowed(UserModel user){
+        // check client can impersonate user
+        if (!AdminPermissions.management(session, realm).users().canClientImpersonate(client, user)) {
+            event.detail(Details.REASON, "client not allowed to impersonate");
+            event.error(Errors.ACCESS_DENIED);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.ACCESS_DENIED, "Client not allowed to perform assertion grant", Response.Status.FORBIDDEN);
+        }
+    }
+
+    /**
+     * Validate the encoded assertion JWT against the cross-domain trusted issuers.
+     * Performs the following checks:
+     *      - ensure JWT is active
+     *      - validate the signature against trusted issuers
+     *      - ensure the issuer and audience in the token match the specified issuer and audience in the trusted issuer configuration that validated
+     *        the JWT signature.
+     * This function throws an HTTP forbidden error if the assertion is not valid
+     *
+     * @param assertion The encoded and signed assertion JWT
+     * @return The parsed assertion JWT
+     */
+    private JsonWebToken validateAssertion(String assertion){
+        JsonWebToken jws = null;
+
+        // set in the verification loop
+        boolean valid = false;
+        List<String> validationErrors = new ArrayList<>();
+
+        if (trustedIssuers.isEmpty()){
+            validationErrors.add(String.format("Client %s has no configured trusted issuers", client.getClientId()));
+        }
+
+        // loop through each config to see if it can verify the assertion
+        for (CrossDomainTrust config: trustedIssuers) {
+            try {
+                // verify the token signature and check the issuer and audience
+                TokenVerifier<JsonWebToken> verifier = TokenVerifier.create(assertion, JsonWebToken.class)
+                        .withChecks(
+                                TokenVerifier.IS_ACTIVE,
+                                new TokenVerifier.AudienceCheck(config.getAudience()),
```

Comment:
I tried to find discussions around the audience constraints but I did not find any.

I would like to know why not allow the token endpoint as a valid audience, even if not defined in the configuration. Perhaps, we could even remove the audience from the configuration and always expect the audience to be the token endpoint?

I think having the token endpoint as a valid audience helps to simplify how assertions are issued because the token endpoint is exposed in the discovery document, making it easier to find the right value to use as the audience and using a standard claim in that document. From a management perspective, it should also help to avoid having to specify audiences because the token endpoint is the same across different combinations of `issuer` and `certificate`.